### ISSUE TITLE
 Bespoke validation functions returning TIMEDOUT error unexpectedly

### ISSUE THREAD

**Issue by LukeLongworth** [maths/moodle-qtype_stack#1121]
Problem: I am getting a TIMEDOUT error with my bespoke validation function for an unknown reason. To try and diagnose it, I recreated the error in a very simple question:
![image](https://github.com/maths/moodle-qtype_stack/assets/34358809/5390c227-cc0f-4985-8d0e-9b4eb565106c)
![image](https://github.com/maths/moodle-qtype_stack/assets/34358809/d02b8690-b950-4bfe-aea8-a16722c0a00b)
Then at the PRT I just check whether ans1 = 1. 

In this form, the question works exactly as intended, but when I un-comment the line that calls an integral, I get a TIMEDOUT error in the validation. This feels very strange to me. The _only_ way I can recreate this error is by calling the defint function to integrate an exponential function with an argument of the form (constant*f(x)). I'm not even assigning values to a variable, I'm simply calling the function in the Question variables

Things that cause this to break:
- `int(e^(s*x),x,0,1)`
- `int(%e^(s*x),x,0,1)`
- `defint(e^(s*x),x,0,1)`
- `int(%e^(s*x^2),x,0,1)`
- `(s:1, int(e^(s*x),x,0,1)`
- `int(e^(m*x),x,0,1)`
- `int(e^(2*x),x,0,1)`
- `laplace(e^(2*x),x,s)`
- `assume(s>0); int(e^(x*ln(s)),x,0,1)`

Things that **don't** cause this to break:
- `int(e^(s*x),x)`
- `int(e^x,x,0,1)`
- `laplace(e^x,x,s)`
- `int(s*x^2,x,0,1)`
- `assume(s>0); int(ln(s*x),x,0,1)`
- `assume(s>0); int(s^x,x,0,1)`
- `ev(subst([x=1],int(e^(s*x),x)) - subst([x=0],int(e^(s*x),x)),simp)`
- `int(exp(s*x),x,0,1)` (this confuses me the most)
- `assume(s>0); int(exp(x*ln(s)),x,0,1)`

These issues do not persist when I call the problem function somewhere outside of the Question variables (e.g. in the Question text or General Feedback). I am not sure where to find the source code that calls bespoke validation functions, so I don't have any suggestions as to where the issue arises except that it seems to be about the `defint` function, sorry!

**Comment by sangwinc** [comment]
@LukeLongworth I'm very sorry but I can't reproduce this on my test server!  Can you email me your question and I'll try to break one of the older (live) servers.

Just to note that Maxima has an interactive integrator.  Try the following on a desktop Maxima system

integrate(x^n,x);

It is possible this interactivity is causing the timeout.

**Comment by LukeLongworth** [comment]
I've returned to this issue and have some updates: 

Firstly, it doesn't seem to be `defint` after all, as I've discovered some extra functions that cause the issue: 
- `int((sin(t)*e^(-2*t),t)`
- `int((sinh(t)*e^(-2*t),t)`
- `int((ln(t)*e^(-2*t),t)`
Even though the following do _not_ cause the issue:
- `int(((e^(i*t) - e^(-i*t))/(2*i))*e^(-2*t),t)`
- `int(((e^(t) - e^(-t))/2)*e(-2*t),t)`

Secondly, this issue is persisting in STACK version 2024012900 in a different installation. Santiago kindly added me to an IDEMS test server so I could check, and I am observing the exact same behaviour, both when I import an existing question and create a new one. I've included this question (imported once from each server) below. I observe this this on both my work device and on my phone.

[custom validation glitch.zip.zip](https://github.com/maths/moodle-qtype_stack/files/14951594/custom.validation.glitch.zip.zip)

**Comment by sangwinc** [comment]
@LukeLongworth Since I can't reproduce this, and we've released a new version since it was opened I'm closing it.  Please let me know if this is still an issue and I'll help fix it.

**Comment by LukeLongworth** [comment]
@sangwinc I think the issue is still ongoing, but we only upgrade annually. If this is still an issue once we upgrade in January I'll let you know. 

I didn't hear back from our IT team about "timeout" in the STACK settings. I'll chase that up and let you know what they say.

**Comment by sangwinc** [comment]
@LukeLongworth could you confirm if this is still and issue?  I'd like to fix this before Nov 19th (next release) or close it!

**Comment by sangwinc** [comment]
@LukeLongworth I've just tried your example on my development setup with the release candidate code, and can't reproduce the problem. I'm going to bump this to the next release.

**Comment by LukeLongworth** [comment]
Kia ora @sangwinc, I've just tried this again on version 4.7.0 and it's still persisting. I don't have the required access to check settings in the install, but I'm meeting with IT this afternoon and will bring this up then if we have time. Thank you for continuing to check this!

**Comment by EJMFarrow** [comment]
I managed to replicate this locally by including two of the offending functions in question variables. (Each on its own is fine.)
```
a:int(e^(s*x),x,0,1);
b:int(ln(t)*e^(-2*t),t);

validate_test(ex):= block(
   return("")
);
```
Oddly, upping the CAS timeout doesn't seem to fix it. I'll investigate further.

**Comment by EJMFarrow** [comment]
This appears to be to do with simplification. If you have a question with e.g. `c:int(e^(s*x),x,0,1);` in question variables, set 'question-level simplify to 'No' and attempt to save you will immediately get an error `Condition in MACSYMA-TOP-LEVEL [or a callee]: INTERNAL-SIMPLE-ERROR: Value stack overflow.`

By default, validation does not simplify and maxima falls over when it tries to process the question variables as part of the custom validation. A workaround is to include `simp` in the answer extra options but this may not be exactly what you want. @sangwinc - any thoughts?

**Comment by sangwinc** [comment]
Here is a minimal example, in any maxima desktop session.

````
simp:false;
integrate(log(t)*e^(-2*t),t);
````

So, it's not really anything to do with STACK itself.

**Comment by sangwinc** [comment]
Yes, I think the solution to the problem is to surround those integrals with `ev(.... ,simp)` to ensure they are always simplified.  Let me know if this doesn't work and I'll look again!
